### ISSUE TITLE
SDK 2.0 compatibility

### ISSUE THREAD

**Issue by tjkirch** [pd-rs/crankstart#45]
Panic sent out [this email](https://lists.play.date/w/a8l763VN6Cf4e2f2hWjpPP5g) about the upcoming SDK 2.0 update, which links to [this detail page](https://help.play.date/developer/os-2.0/).

I tested my simple crankstart-based game with SDK 2.0 - it ran fine in the simulator, but it crashes on my Playdate, saying `CAPI handler function wasn't located in loaded data`.  I'm wondering what's required for compatibility.

Here are some random thoughts:
* [Build requirements](https://help.play.date/developer/os-2.0/#build-requirements): `pdc` now needs to know about ELF relocations via a pdex.elf file.  I don't see it generating one automatically.  I'm not knowledgeable in this area; is this a concern, or is Rust doing automatically what Panic suggests in their updated compiler/linker flags?  I saw someone mention `emit-relocs` [for tock](https://github.com/tock/libtock-rs/issues/28#issuecomment-415179432).
* Similar for updated [linker requirements](https://help.play.date/developer/os-2.0/#linker-requirements).
* Once those basics are figured out: SDK 2.0 games aren't compatible with 1.x Playdates, and SDK 1.x games aren't compatible with 2.0 Playdates.  Can it just be up to the crankstart user to pick an SDK and crankstart will work (possibly via detection) or will there need to be a flag?  Maybe 1.x compatibility just won't matter?

**Comment by boozook** [comment]
Same Issue for me. We're unable to test on the device because latest released OS version is 1.13.7 without new loader. We just can hope :))
[official note](https://help.play.date/developer/os-2.0/#-publish-your-20-build)
They will change elf loader in the future OS 2.0.  

> Playdates on OS 1.x will not be able to launch your 2.0 build, but will display a dialog asking the user to upgrade their Playdate’s OS. __If OS 2.0 is not available to the public, the user will not be able to run your game.__

I'll try to get "normal execution" of new build (for new loader with relloc support) on the rev1 device with OS 1.13.7 on board to get message like "The Playdate OS is too old to run this game".
I'm working on it...

**Comment by boozook** [comment]
Update.
Actually nothing interesting. I've build a correct binary (as I suppose) with rust and rust-lld, but after `pdc` there's nothing (only 130 bytes of nothing).
<details> 
  <summary>My test "foo" binary contains the following...</summary>

I'm sure it should work because now in linker-map `eventHandlerShim` as `ENTRY` explicitly.
And I have one exactly as needed, same as in the setup.c.

```text
__bss_end__
__bss_start__
__data_end__
__data_start__
__etext
eventHandlerShim

foo:	file format elf32-littlearm

Program Header:
    LOAD off    0x00010000 vaddr 0x00000000 paddr 0x00000000 align 2**16
         filesz 0x00001774 memsz 0x00001774 flags r-x
    LOAD off    0x00011774 vaddr 0x00001774 paddr 0x00001774 align 2**16
         filesz 0x00000000 memsz 0x00000004 flags rw-
   STACK off    0x00000000 vaddr 0x00000000 paddr 0x00000000 align 2**64
         filesz 0x00000000 memsz 0x00000000 flags rw-

Dynamic Section:

Sections:
Idx Name                Size     VMA      LMA      Type
  0                     00000000 00000000 00000000 
  1 .text               00001774 00000000 00000000 TEXT
  2 .data               00000000 00001774 00001774 TEXT
  3 .bss                00000004 00001774 00001774 BSS
  4 .rel.text           00000598 00000000 00000000 
  5 .debug_abbrev       000003ff 00000000 00000000 DEBUG
  6 .debug_info         00002b59 00000000 00000000 DEBUG
  7 .rel.debug_info     000026a8 00000000 00000000 
  8 .debug_aranges      000001f0 00000000 00000000 DEBUG
  9 .rel.debug_aranges  000001a0 00000000 00000000 
 10 .debug_ranges       00001028 00000000 00000000 DEBUG
 11 .rel.debug_ranges   000017f0 00000000 00000000 
 12 .debug_str          000067a0 00000000 00000000 DEBUG
 13 .debug_pubnames     0000255b 00000000 00000000 DEBUG
 14 .rel.debug_pubnames 00000028 00000000 00000000 
 15 .debug_pubtypes     0000005a 00000000 00000000 DEBUG
 16 .rel.debug_pubtypes 00000028 00000000 00000000 
 17 .ARM.attributes     00000047 00000000 00000000 
 18 .debug_frame        00000718 00000000 00000000 DEBUG
 19 .rel.debug_frame    00000430 00000000 00000000 
 20 .debug_line         00002580 00000000 00000000 DEBUG
 21 .rel.debug_line     00000178 00000000 00000000 
 22 .comment            00000013 00000000 00000000 
 23 .symtab             00000830 00000000 00000000 
 24 .shstrtab           00000148 00000000 00000000 
 25 .strtab             00000f70 00000000 00000000 
```

🤷‍♂️

</details>

**Comment by boozook** [comment]
One more update. I've compiled one of examples in the SDK/C_API/Examples...
Just to compare resulting elf. And you know what? When I run it on the device, I see "no capi handler"!
Well, I'm not sure that its possible right now if I can't run example built with pdc2.0-beta on my hardware with OSv1.13.7.
All ideas are welcome.

**Comment by tjkirch** [comment]
> One more update. I've compiled one of examples in the SDK/C_API/Examples... Just to compare sections in the resulting elf. And you know what? When I run it on the device, I see "no capi handler"! Well, I'm not sure that its possible right now if I can't run example built with pdc2.0-beta on my hardware with OSv1.13.7. All ideas are welcome.

@boozook Interesting - I was able to get the expected "Update Needed" message on my 1.13.7 device.  I just unpacked the 2.0 beta2 SDK to my `$PLAYDATE_SDK_PATH`, did `cd $PLAYDATE_SDK_PATH/C_API/Examples/Hello World`, then `make all`, and copied `HelloWorld.pdx` to my Playdate.  Is there any chance you had the path variable pointing to a 1.x SDK, or something like that?

Two things surprised me, though - the pdx had no ELF relocations file, and it ran on both 1.x and 2.x simulators, which I didn't think was supposed to work.  I'd have to guess that the simulator is special-cased to run 1.x and 2.x games.

**Comment by boozook** [comment]
@tjkirch I've reported about it on the dev forum. There's video with reproduction. [Link](https://devforum.play.date/t/sdk-2-0-b2-pdc-produces-pix-with-broken-binary/11345)

**Comment by tjkirch** [comment]
@boozook Do you think there's any way others could help with your investigation?  I'm not experienced in this area, but I'm worried about the SDK 2 release obsoleting all crankstart games, so I'd like to do what I can!

**Comment by tjkirch** [comment]
Panic just announced that SDK 2.0 is going live in a few days, on June 12th, 2023. They make it easy to update, so most users will probably do so quickly, and then crankstart games will no longer work. We definitely need to figure this out.

**Comment by rtsuk** [comment]
Has anyone shipped a game using crankstart?

**Comment by tjkirch** [comment]
> Has anyone shipped a game using crankstart?

I'm not sure, but I'd like to. I've spent a long time on mine, and I'd be quite sad to have to rebuild it in Lua.

**Comment by boozook** [comment]
I have an MVP-like build script "rust build executable bin then give it to pdc". I absolutely not sure that it works, so I just need to test it with latest beta. Or better test on real hardware, I have one, but so have to wait OS 2.0 release. 🤷🏻‍♂️

**Comment by boozook** [comment]
@tjkirch thank you, I didn’t know that.
> Playdate OS/SDK Beta Tester Sign Up
> Your response has been recorded

**Comment by boozook** [comment]
I'm working on it... 😐

**Comment by scratchminer** [comment]
Here's what

**Comment by boozook** [comment]
Sad news for now. I tried really many options and didn't understand where I was wrong. So I have no idea for now.

[There is dummy isolated example (/playground).](https://github.com/boozook/pdRevB-problem)

@scratchminer, @tjkirch, do you have any ideas. Seems to I need help.

**Comment by boozook** [comment]
Hm... Probably I have another option. Tomorrow I'll try to use __only__ gcc linking. Will see...

**Comment by boozook** [comment]
Okay. Almost all my previous solutions did not works because of bug in the pdc.
Now building works fine. Testing relocations and fix will be here soon.

**Comment by boozook** [comment]
It coulee really useful as "hot-reload" feature for develop/debug in sim only. I'll think about it after revB & OS2.0 support.

**Comment by boozook** [comment]
Status update:
- Simple build lib with rustc+arm-gcc works
- I have some trouble with alloc, just strange cstrings behaviour, solvable.
Sims I close to finish.

**Comment by boozook** [comment]
Probably I'm done.
Please test and review.

**Comment by boozook** [comment]
Also we're need some validity tests for produced output to detect regression on future changes.

**Comment by boozook** [comment]
I need feedback to merge [crank#42](https://github.com/pd-rs/crank/pull/42).

**Comment by boozook** [comment]
I would close this because there's a latest fix in #48.
